File: cms/server/pkg/integrationapi/comment.go
package integrationapi

import (
	"github.com/reearth/reearth-cms/server/pkg/thread"
	"github.com/samber/lo"
)

func NewComment(c *thread.Comment) *Comment {
	if c == nil {
		return nil
	}

	var authorID any
	var authorType CommentAuthorType
	if c.Author().User() != nil {
		authorID = c.Author().User().Ref()
		authorType = User
	}
	if c.Author().Integration() != nil {
		authorID = c.Author().Integration().Ref()
		authorType = Integrtaion
	}
	return &Comment{
		Id:         c.ID().Ref(),
		AuthorId:   &authorID,
		AuthorType: &authorType,
		Content:    lo.ToPtr(c.Content()),
		CreatedAt:  lo.ToPtr(c.CreatedAt()),
	}
}

File: cms/web/src/gql/queries/comment.ts
import { gql } from "@apollo/client";

export const ADD_COMMENT = gql`
  mutation AddComment($threadId: ID!, $content: String!) {
    addComment(input: { threadId: $threadId, content: $content }) {
      comment {
        id
        author {
          ... on User {
            id
            name
            email
          }
          ... on Integration {
            id
            name
          }
        }
        authorType
        authorId
        content
        createdAt
      }
    }
  }
`;

export const UPDATE_COMMENT = gql`
  mutation UpdateComment($commentId: ID!, $threadId: ID!, $content: String!) {
    updateComment(input: { commentId: $commentId, threadId: $threadId, content: $content }) {
      comment {
        id
        author {
          ... on User {
            id
            name
            email
          }
          ... on Integration {
            id
            name
          }
        }
        authorType
        authorId
        content
        createdAt
      }
    }
  }
`;

export const DELETE_COMMENT = gql`
  mutation DeleteComment($commentId: ID!, $threadId: ID!) {
    deleteComment(input: { commentId: $commentId, threadId: $threadId }) {
      commentId
    }
  }
`;
